File: app/src/main/java/com/sakurafish/pockettoushituryou/di/ApplicationModule.java
package com.sakurafish.pockettoushituryou.di;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.sakurafish.pockettoushituryou.data.db.entity.OrmaDatabase;
import com.squareup.moshi.Moshi;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.disposables.CompositeDisposable;

@Module
public class ApplicationModule {
    private final Application application;

    public ApplicationModule(Application application) {
        this.application = application;
    }

    @Provides
    @Singleton
    Context provideApplicationContext() {
        return application.getApplicationContext();
    }

    @Provides
    public SharedPreferences provideSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    public CompositeDisposable provideCompositeDisposable() {
        return new CompositeDisposable();
    }

    @Singleton
    @Provides
    public OrmaDatabase provideOrmaDatabase(Context context) {
        return OrmaDatabase.builder(context).trace(false).build();
    }

    @Provides
    public Moshi provideMoshi() {
        return new Moshi.Builder().build();
    }
}

File: app/src/main/java/com/sakurafish/pockettoushituryou/di/ApplicationModule.kt
package com.sakurafish.pockettoushituryou.di

import android.app.Application
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import com.sakurafish.pockettoushituryou.data.db.entity.orma.OrmaDatabase
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import dagger.Module
import dagger.Provides
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Singleton

@Module
class ApplicationModule(val application: Application) {

    @Provides
    fun provideSharedPreferences(): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(application)
    }

    @Provides
    fun provideCompositeDisposable(): CompositeDisposable {
        return CompositeDisposable()
    }

    @Singleton
    @Provides
    fun provideOrmaDatabase(): OrmaDatabase {
        return OrmaDatabase.builder(application).trace(false).build()
    }

    @Provides
    fun provideMoshi(): Moshi {
        val moshi = Moshi.Builder()
            .addLast(KotlinJsonAdapterFactory())
            .build()
        return moshi
    }
}
